### PR TITLE
- Use C++11 for loops.

### PR DESCRIPTION
**Pull request by baldersheim**
- Detangle some includes and include what you need.
- Refactor to subclass to avoid code duplication.
- Add config for compression too.
- V1 up to 142, V2 above.
@havardpe and @vekterli PR

### CODE REVIEW

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        p.retry = (int)root.field(RETRY_F).asLong();
+        p.timeRemaining = root.field(TIMEREMAINING_F).asLong();
+        p.protocolName = new Utf8Array(Utf8.toBytes(root.field(PROTOCOL_F).asString()));
+        p.payload = root.field(BLOB_F).asData();
+        p.traceLevel = (int)root.field(TRACELEVEL_F).asLong();
+        return p;
+    }
+
+    @Override
+    protected void createReponse(Values ret, Reply reply, Version version, byte [] payload) {
```
Comment (vekterli) [inline messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java]: Typo: Reponse -> Response

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        root.setBool(USERETRY_F, msg.getRetryEnabled());
+        root.setLong(RETRY_F, msg.getRetry());
+        root.setLong(TIMEREMAINING_F, msg.getTimeRemaining());
+        root.setLong(TRACELEVEL_F, traceLevel);
+        root.setData(BLOB_F, payload);
+
+        byte[] serializedSlime = BinaryFormat.encode(slime);
+        Compressor.Compression compressionResult = compressor.compress(serializedSlime);
+        Request req = new Request(METHOD_NAME);
+        Values v = req.parameters();
```
Comment (vekterli) [inline messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java]: Assuming we want to move to TLS and to be able to extend request headers with secret-bearing Authorization headers etc in the future, compressing these together with an arbitrary payload is unfortunately not secure. See https://en.wikipedia.org/wiki/CRIME and https://en.wikipedia.org/wiki/BREACH.

tl;dr assuming an adversary can observe encrypted packets and cause the system to send packets containing own data (user feed/queries is a very realistic, non-theoretical case), the resulting compressed length of the ciphertext is used as an oracle that allows you to iteratively deduce  secret contents in _O(n)_ rather than _O(2^n)_.

Separating headers from payload and only compressing the latter will at least ensure we don't leak our own secrets.

File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+class ParamsV2 : public RPCSend::Params
+{
+public:
+    ParamsV2(const FRT_Values &arg)
+        : _slime()
+    {
+        uint8_t encoding = arg[0]._intval8;
+        uint32_t uncompressedSize = arg[1]._intval32;
+        DataBuffer uncompressed(arg[2]._data._buf, arg[2]._data._len);
+        ConstBufferRef blob(arg[2]._data._buf, arg[2]._data._len);
```
Comment (vekterli) [inline messagebus/src/vespa/messagebus/network/rpcsendv2.cpp]: Dangerous to trust buffer lengths received from the network. Should check lengths against received blob size.

File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+std::unique_ptr<Reply>
+RPCSendV2::createReply(const FRT_Values & ret, const string & serviceName,
+                       Error & error, vespalib::TraceNode & rootTrace) const
+{
+    uint8_t encoding = ret[0]._intval8;
+    uint32_t uncompressedSize = ret[1]._intval32;
+    DataBuffer uncompressed(ret[2]._data._buf, ret[2]._data._len);
+    ConstBufferRef blob(ret[2]._data._buf, ret[2]._data._len);
```
Comment (vekterli) [inline messagebus/src/vespa/messagebus/network/rpcsendv2.cpp]: Dangerous to trust buffer lengths received from the network. Should check lengths against received blob size.

**Review by vekterli — COMMENTED** [review]
Some comments from first pass through the diff. Need more time for in-depth look.

File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+Memory BLOB_F("msg");
+Memory ERRORS_F("errors");
+Memory CODE_F("code");
+Memory MSG_F("msg");
+Memory SERVICE_F("service");
+
+
+}
+
+bool RPCSendV2::isCompatible(stringref method, stringref request, stringref respons)
```
Comment (vekterli) [inline messagebus/src/vespa/messagebus/network/rpcsendv2.cpp]: Typo: respons

File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp (hunk 4 above)
Comment (vekterli) [reply messagebus/src/vespa/messagebus/network/rpcsendv2.cpp]: Ignore comment, did not originally see that length is from RPC layer itself 🦄 👍

File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp (hunk 3 above)
Comment (vekterli) [reply messagebus/src/vespa/messagebus/network/rpcsendv2.cpp]: Ignore comment, did not originally see that length is from RPC layer itself 🦄 👌

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java (hunk 1 above)
Comment (baldersheim) [reply messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java]: Fixed

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+import com.yahoo.messagebus.routing.Route;
+import com.yahoo.slime.BinaryFormat;
+import com.yahoo.slime.Cursor;
+import com.yahoo.slime.Inspector;
+import com.yahoo.slime.Slime;
+import com.yahoo.text.Utf8;
+import com.yahoo.text.Utf8Array;
+
+/**
+ * Implements the request adapter for method "mbus.sendslime".
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java]: Fixed

File: messagebus/src/vespa/messagebus/network/rpcnetwork.cpp [hunk 7]
```diff
@@ -104,24 +107,26 @@ RPCNetwork::TargetPoolTask::PerformTask()
...
-    _targetPool(params.getConnectionExpireSecs()),
-    _targetPoolTask(_scheduler, _targetPool),
-    _servicePool(*this, 4096),
-    _slobrokCfgFactory(params.getSlobrokConfig()),
-    _mirror(std::make_unique<slobrok::api::MirrorAPI>(_orb, _slobrokCfgFactory)),
-    _regAPI(std::make_unique<slobrok::api::RegisterAPI>(_orb, _slobrokCfgFactory)),
-    _oosManager(_orb, *_mirror, params.getOOSServerPattern()),
+    _threadPool(std::make_unique<FastOS_ThreadPool>(128000, 0)),
+    _transport(std::make_unique<FNET_Transport>()),
+    _orb(std::make_unique<FRT_Supervisor>(_transport.get(), _threadPool.get())),
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcnetwork.cpp]: I wanted to control the threadpool on the outside as I assumed that one would be create if it was NULL. Looking at the code I saw that none was created, and then some methods would not work. Hnece I thought it more consistent to give it a threadpool.

File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp (hunk 5 above)
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcsendv2.cpp]: Fixed

File: messagebus/src/vespa/messagebus/network/rpcnetwork.h [hunk 8]
```diff
@@ -51,24 +58,26 @@ class RPCNetwork : public INetwork,
...
+    std::unique_ptr<slobrok::ConfiguratorFactory> _slobrokCfgFactory;
+    std::unique_ptr<slobrok::api::IMirrorAPI>     _mirror;
+    std::unique_ptr<slobrok::api::RegisterAPI>    _regAPI;
+    std::unique_ptr<OOSManager>                   _oosManager;
     int                       _requestedPort;
-    RPCSendV1                 _sendV1;
+    std::unique_ptr<RPCSendAdapter> _sendV1;
+    std::unique_ptr<RPCSendAdapter> _sendV2;
     SendAdapterMap            _sendAdapters;
+    CompressionConfig         _compressionConfig;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcnetwork.h]: Fixed

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        v.setData(name, vespalib::Memory(_payload.data(), _payload.size()));
+    }
+private:
+    mutable Blob _payload;
+};
+
+}
+
+RPCSend::RPCSend() :
+    _net(NULL),
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcsend.cpp]: Fixed

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java (hunk 2 above)
Comment (baldersheim) [reply messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java]: use body, and allow for later header payload.

File: messagebus/src/vespa/messagebus/network/rpcnetwork.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (havardpe) [reply messagebus/src/vespa/messagebus/network/rpcnetwork.cpp]: This is kind of deprecated functionality for non-instant invocations (the supervisor using the thread pool). We always want to use detached instant requests instead.

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+private:
+    BlobRef _payload;
+};
+
+class FillByHandover final : public PayLoadFiller
+{
+public:
+    FillByHandover(Blob payload) : _payload(std::move(payload)) { }
+    void fill(FRT_Values & v) const override {
+        v.AddData(std::move(_payload.payload()), _payload.size());
```
Comment (havardpe) [inline messagebus/src/vespa/messagebus/network/rpcsend.cpp]: this feels bad. Please sample size before moving

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    }
+    void fill(const vespalib::Memory & name, vespalib::slime::Cursor & v) const override {
+        v.setData(name, vespalib::Memory(_payload.data(), _payload.size()));
```
Comment (havardpe) [inline messagebus/src/vespa/messagebus/network/rpcsend.cpp]: I expect the new external memory slime data value will be used here?

Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcsend.cpp]: Correct, here we add the serialized blob from the protocol (documentapi) as an opaque blob.

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp (hunk 10 above)
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcsend.cpp]: Fixed.

Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcsend.cpp]: Ok, so you prefer we leave it as nullptr, and let it err if misused. I agree with that.
